packages/openneuro-app/src/scripts/dataset/download/download-native.js

```diff
@@ -40,30 +40,90 @@ class DownloadAbortError extends Error {
   }
 }
 
+let downloadCanceled
+
 /**
- * Downloads a dataset via the native file API, skipping expensive compression if the browser supports it
- * @param {string} datasetId Accession number string for a dataset
- * @param {string} snapshotTag Snapshot tag name
+ * Recursive download for file trees via browser file access API
  */
-export const downloadNative = (datasetId, snapshotTag, client) => async () => {
+const downloadTree = async (
+  { datasetId, snapshotTag, client, apmTransaction, dirHandle, toastId },
+  path = '',
+  tree = null,
+) => {
   const filesToDownload = await downloadDataset(client)({
     datasetId,
     snapshotTag,
+    tree,
   })
+  for (const [index, file] of filesToDownload.entries()) {
+    const downloadPath = path ? `${path}/${file.filename}` : file.filename
+    if (file.directory) {
+      // Next tree level
+      await downloadTree(
+        {
+          datasetId,
+          snapshotTag,
+          client,
+          apmTransaction,
+          dirHandle,
+          toastId,
+        },
+        downloadPath,
+        file.id,
+      )
+    } else {
+      // Regular file
+      if (downloadCanceled) {
+        throw new DownloadAbortError('Download canceled by user request')
+      }
+      const fileHandle = await openFileTree(
+        dirHandle,
+        path ? `${path}/${file.filename}` : file.filename,
+      )
+      // Skip files which are already complete
+      if (fileHandle.size == file.size) continue
+      const writable = await fileHandle.createWritable()
+      const { body, status, statusText } = await fetch(file.urls.pop())
+      let loaded = 0
+      const progress = new TransformStream({
+        transform(chunk, controller) {
+          downloadToastUpdate(toastId, loaded / file.size, {
+            datasetId,
+            snapshotTag,
+            downloadPath,
+            dirName: dirHandle.name,
+          })
+          loaded += chunk.length
+          controller.enqueue(chunk)
+        },
+      })
+      if (status === 200) {
+        await body.pipeThrough(progress).pipeTo(writable)
+      } else {
+        apmTransaction.captureError(statusText)
+        return requestFailureToast()
```

Comment:
That's right. You can cancel the entire download by closing the toast but otherwise I think we want to try to get as much as possible, so that if something is missed you can resume with a second attempt and only need a small subset to get a complete dataset.

That brings up a fix here though, I think this should show you if any files failed due to network or other interruptions at the end and suggest a retry. Otherwise it's easy to assume you have the whole dataset when one file is missing.